utils/reusable_buffer.hh

```diff
@@ -8,123 +8,156 @@
+ * sharing the same buffer, if you are not careful.
+ *
+ * To make this buffer slightly harder to misuse, it's only accessible
+ * by reusable_buffer_guard.
+ * There can only exist one guard at a time.
+ * The guard can be used to obtain a view at most once.
+ */
+template <typename Clock>
+class reusable_buffer {
+    friend reusable_buffer_guard<Clock>;
+public:
+    using decay_period = typename Clock::duration;
+private:
+    size_t _high_watermark = 0;          // Max size since the last decay().
+    std::vector<bytes::value_type> _buf; // The underlying contiguous buffer.
+    decay_period _period;                // The period of decay().
+    seastar::timer<Clock> _timer;        // Calls decay() with _period.
+    size_t _refcount = 0;                // Used purely for anti-misuse checks.
+    size_t _reallocs = 0;                // Number of size changes.
 private:
-    bytes_mutable_view reserve(size_t n) {
-        if (_size < n) {
-            // Reusable buffers are expected to be used when large contiguous
-            // allocations are unavoidable. There is not much point in warning
-            // about them since there isn't much that can be done.
-            seastar::memory::scoped_large_allocation_warning_disable g;
-            // std::make_unique would zero-initialise the buffer which is
-            // just a waste of cycles. We can, however, summon an ancient
-            // entity from the elder days of C++ to help us.
-            _buffer.reset(new int8_t[n]);
-            _size = n;
+    void resize(size_t new_size) & {
+        // Reusable buffers are expected to be used when large contiguous
+        // allocations are unavoidable. There is no point in warning
+        // about them since there isn't much that can be done.
+        seastar::memory::scoped_large_allocation_warning_disable g;
+
+        // Clear before shrinking so that the old buffer
+        // is freed before allocating the new buffer.
+        // This lesses the pressure on the allocator and should guarantee
+        // a success if the new size is smaller than the old size.
+        _buf.clear();
+        _reallocs += 1;
+        _buf.resize(new_size);
+    }
+
+    void decay() & {
+        assert(_refcount == 0);
+        if (_high_watermark <= _buf.size() / 16) {
+            // We shrink when the size falls at least by four power-of-2
+            // notches, instead of just one notch. This adds hysteresis:
+            // it prevents small oscillations from inducing shrink/grow cycles.
+            // With the factor of 16 above, only outliers at least 8x bigger than
+            // the otherwise stable size might cause reallocations.
+            //
+            // In other words, we "define" a pathological request as one that is
+            // at least 8x bigger than the stable size, where the stable size
+            // is the max of all sizes seen within a decay period.
+            resize(_high_watermark);
```

Comment:
I think we should shrink to 2*_high_watermark here. The logic is that is (say) the typical request size hovers round 256k - epsilon, we'd allocate 512k, not 256k, and catch small deviations.

Better to waste some memory than to cause unnecessary large allocations. They are very expensive!

I'd also raise the period to 10 minutes (no real justification).

Reply:
> The logic is that is (say) the typical request size hovers round 256k - epsilon, we'd allocate 512k, not 256k, and catch small deviations.

Disagreed. The resistance to small deviations is already assured by the surrounding `if`: we shrink only if we haven't observed in the last period any allocations at least 1/16 as big as the current size. If a small oscillation (around a single power of 2) of the running max is happening, two large allocations may happen (iff the first allocation is smaller than the power of 2), but no more.

Rounding the size up to the *second* greater-or-equal power of 2 could seem to guarantee that *exactly* one allocation happens, instead of at most two. But this isn't true, because the initial allocations may be *2x smaller* than the running max, so they will cause the same reallocation anyway. So it only adds complexity without additional benefit.

> I'd also raise the period to 10 minutes (no real justification).

Ok, why not.